strings/base_collections_base.h

```diff
@@ -1,3 +1,43 @@
+namespace winrt::impl
+{
+    template <typename D, typename = void>
+    struct has_exclusive_op : std::false_type {};
+
+    template <typename D>
+    struct has_exclusive_op<D, std::void_t<decltype(std::declval<D*>()->exclusive_op(std::true_type{}))>> : std::true_type {};
+
+    template <typename D, typename = void>
+    struct has_shared_op : std::false_type {};
+
+    template <typename D>
+    struct has_shared_op<D, std::void_t<decltype(std::declval<D*>()->shared_op(std::true_type{}))>> : std::true_type {};
+
+    struct single_threaded_collection_base
+    {
```

Comment:
Seems like some code would be simplified if we removed the `has_exclusive_op` and `has_shared_op` traits, and the checks for said traits, and simply added `exclusive_op` and `shared_op` to `single_threaded_collection_base`, with these implementations simply calling the op.

Reply:
The priority for this PR should be:

1. Correctness
2. Simplicity of implementation
3. Performance

No one in their right mind is going to use a COM-based collection as a high-performance concurrent collection. This is about correctness in the face of accidental or intentional misuse. Obviously it needs to be correct, but beyond that we need to strive for the simplest practical solution as complexity here can really make this difficult to maintain.

Reply:
That's actually how I had it initially. The problem there is that these types are not in the `impl` namespace and I wanted to avoid a breaking change. In general, the following two approaches were considered:

OPTION 1: Have the `perform_exclusive` function in `iterable_base` just call the function object and then have `perform_shared` do the forwarding call to `perform_exclusive`, and then the implementation class can provide alternative definition(s) of those two functions. This is actually why there's still a bunch of CRTP casting done when calling `perform_*` that I neglected to remove. Oops. The reason I abandoned that route was because it made the implementations more difficult to write. I wanted the `*_threaded_collection_base` types that I could derive from, but then you'd have two base classes that define the same functions, which would require awkward `using Base::perform_*` statements. Overall, it's probably not that bad - at the time it was a little bit more awkward based on the way things were structured at the time. Now that the single/multi-threaded definitions are just usings, it's probably not a huge deal any more.

OPTION 2: Take the breaking change and add `ThreadingBase` as a template argument that persists all the way down to `iterable_base`. This is actually more-or-less required for doing per-iterator locking should we decide we want that. It's a pretty minor breaking change, so perhaps we don't care all that much.

Reply:
The latest commit basically does option 1 from above. It's slightly awkward since both `single_threaded_collection_base` _and_ `iterable_base` define the same thing for `perform_shared`/`perform_exclusive`, but that's just a side effect of `using ThreadingBase::perform_*` needing to reference something that exists.

Reply:
I was expecting something more like

```cpp
struct single_threaded_collection_base
{
    struct nop {};
    auto with_lock_guard() { return nop{}; }
    auto with_shared_lock_guard() { return nop{}; }
};
struct multi_threaded_collection_base
{
    auto with_lock_guard() { return lock_guard(m_mutex); }
    auto with_shared_lock_guard() { return shared_lock_guard(m_mutex); }
private:
   srwlock m_mutex;
};
```

and then the classes would do

```cpp
auto guard = this->with_lock_guard();
blah blah blah;
```

or

```cpp
auto guard = this->with_shared_lock_guard();
blah blah blah;
```

Most of the changes would be just dropping `auto guard = ...` into various methods, which makes it easier to review because that preserves existing indentation levels and increases confidence that no funny business snuck in.

The wrapped lambda is more complicated to use because you have to double-return (once to return a value from the lambda, and another to propagate that value out of the function). It also removes reliance on the compiler to inline the lambda.